src/main/java/edu/kit/kastel/highlighter/HighlighterManager.java

```diff
@@ -0,0 +1,282 @@
+import com.intellij.openapi.editor.markup.GutterIconRenderer;
+import com.intellij.openapi.editor.markup.HighlighterLayer;
+import com.intellij.openapi.editor.markup.HighlighterTargetArea;
+import com.intellij.openapi.editor.markup.RangeHighlighter;
+import com.intellij.openapi.editor.markup.TextAttributes;
+import com.intellij.openapi.fileEditor.FileDocumentManager;
+import com.intellij.openapi.fileEditor.FileEditorManager;
+import com.intellij.openapi.fileEditor.FileEditorManagerListener;
+import com.intellij.openapi.fileEditor.TextEditor;
+import com.intellij.openapi.ui.popup.JBPopup;
+import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.ui.AnActionButton;
+import com.intellij.ui.JBColor;
+import edu.kit.kastel.extensions.guis.AnnotationsListPanel;
+import edu.kit.kastel.extensions.settings.ArtemisSettingsState;
+import edu.kit.kastel.icons.ArtemisIcons;
+import edu.kit.kastel.sdq.artemis4j.grading.Annotation;
+import edu.kit.kastel.state.PluginState;
+import edu.kit.kastel.utils.EditorUtil;
+import org.jetbrains.annotations.NotNull;
+
+public class HighlighterManager {
+    private static final Map<Editor, List<HighlighterWithAnnotations>> highlightersPerEditor = new IdentityHashMap<>();
+
+    private static int lastPopupLine;
+    private static Editor lastPopupEditor;
+    private static JBPopup lastPopup;
+
+    public static void initialize() {
+        var messageBus = EditorUtil.getActiveProject().getMessageBus();
+        messageBus.connect().subscribe(FileEditorManagerListener.FILE_EDITOR_MANAGER, new FileEditorManagerListener() {
+            @Override
+            public void fileOpened(@NotNull FileEditorManager source, @NotNull VirtualFile file) {
+                var editor = source.getSelectedTextEditor();
+                if (PluginState.getInstance().isAssessing()) {
+                    editor.getDocument().setReadOnly(true);
+                    updateHighlightersForEditor(editor);
+                }
+            }
+
+            @Override
+            public void fileClosed(@NotNull FileEditorManager source, @NotNull VirtualFile file) {
+                var editor = source.getSelectedTextEditor();
+                if (editor == null) {
+                    return;
+                }
+
+                clearHighlightersForEditor(editor);
+            }
+        });
+
+        PluginState.getInstance().registerAssessmentStartedListener(assessment -> {
+            assessment.registerAnnotationsUpdatedListener(annotations -> updateHighlightersForAllEditors());
+        });
+
+        // When an assessment is closed, clear everything
+        PluginState.getInstance().registerAssessmentClosedListener(() -> {
+            highlightersPerEditor.clear();
+            cancelLastPopup();
+        });
+    }
+
+    public static void onMouseMovedInEditor(EditorMouseEvent e) {
+        // var highlighters = highlightersPerEditor.get(e.getEditor());
```

Comment:
Why is this code disabled?

Reply:
Because I still want to implement something like this later. There are many things like that across the codebase, I just didn't have the time yet to implement everything or create issues for everything